File: backend/src/execution/razorpayExecutor.js
/**
 * razorpayExecutor.js — Real Razorpay test-mode Payment Links.
 *
 * Used for ONLY 2-3 hand-picked demo_case=true cases.
 * This is the ONLY file that converts rupees to paise.
 *
 * IMPORTANT: Never sets status="recovered" directly — only the
 * webhook handler (razorpayWebhook.js) can do that after
 * confirming a payment_link.paid event.
 */

const razorpay = require('../config/razorpayClient');

/**
 * Execute a recovery action via real Razorpay test-mode Payment Link.
 *
 * @param {object} caseData - The case document from MongoDB
 * @param {object} plan - The approved recovery plan
 * @returns {Promise<object>} Updated execution fields + status
 */
async function executeRazorpay(caseData, plan) {
  const now = new Date();
  const recommendation = plan.recommendation;

  // Only CREATE_PAYMENT_LINK and OFFER_DISCOUNT create real payment links
  if (recommendation === 'CREATE_PAYMENT_LINK' || recommendation === 'OFFER_DISCOUNT') {
    try {
      let finalAmount = caseData.amount;

      // Apply discount if this is an approved OFFER_DISCOUNT action
      if (recommendation === 'OFFER_DISCOUNT' && plan.discount_requested_pct > 0) {
        const discountMultiplier = 1 - (plan.discount_requested_pct / 100);
        finalAmount = Math.round(caseData.amount * discountMultiplier);
      }

      // ──────────────────────────────────────────────────────────────
      // AMOUNT CONVENTION: amounts are stored in whole rupees everywhere.
      // Razorpay's API requires paise (subunits). Convert ONLY here:
      const amountInPaise = finalAmount * 100;
      // ──────────────────────────────────────────────────────────────

      const paymentLink = await razorpay.paymentLink.create({
        amount: amountInPaise,
        currency: caseData.currency || 'INR',
        description: `RevivePay Recovery - Case ${caseData._id}`,
        customer: {
          name: `Customer ${caseData.customer_id}`,
          contact: '+919876543210' // Must not have recurring digits for Razorpay test-mode
        },
        notify: {
          sms: false,
          email: false
        },
        reminder_enable: false,
        notes: {
          case_id: caseData._id.toString(),
          original_amount: caseData.amount,
          discount_pct: plan.discount_requested_pct || 0,
          source: 'revivepay'
        }
      });

      return {
        execution: {
          executor_type: 'razorpay',
          action_taken: recommendation,
          result: 'success',
          razorpay_reference: paymentLink.id,
          payment_link_status: 'created',
          timestamp: now
        },
        // NEVER set "recovered" here — only the webhook can do that
        status: 'awaiting_payment',
        contact_count: caseData.contact_count + 1,
        last_contacted_at: now,
        payment_link_url: paymentLink.short_url
      };
    } catch (error) {
      console.error(`[RazorpayExecutor] Error creating payment link for case ${caseData._id}:`, error.error?.description || error.message || error);
      return {
        execution: {
          executor_type: 'razorpay',
          action_taken: recommendation,
          result: 'failure',
          razorpay_reference: null,
          payment_link_status: null,
          timestamp: now
        },
        status: 'execution_failed'
      };
    }
  }

  // SEND_REMINDER — simulate (no real messaging provider)
  if (recommendation === 'SEND_REMINDER') {
    console.log(`[RazorpayExecutor] Simulated reminder for case ${caseData._id}`);
    return {
      execution: {
        executor_type: 'razorpay',
        action_taken: 'SEND_REMINDER',
        result: 'success',
        razorpay_reference: null,
        payment_link_status: null,
        timestamp: now
      },
      status: 'link_created',
      contact_count: caseData.contact_count + 1,
      last_contacted_at: now
    };
  }

  // ESCALATE_TO_HUMAN / DO_NOT_CONTACT — no external call
  return {
    execution: {
      executor_type: 'razorpay',
      action_taken: recommendation,
      result: 'not_attempted',
      razorpay_reference: null,
      payment_link_status: null,
      timestamp: now
    },
    status: recommendation === 'DO_NOT_CONTACT' ? 'stopped_safely' : 'needs_merchant_approval'
  };
}

module.exports = { executeRazorpay };

File: backend/app/execution/razorpay_executor.py
import logging
from datetime import datetime, timezone
from app.models.case import Case, Plan, Execution, ExecutorType, ExecutionResult, PaymentLinkStatus, CaseStatus
from app.config.razorpay_client import razorpay_client

logger = logging.getLogger(__name__)

async def execute_razorpay(case_data: Case, plan: Plan) -> dict:
    now = datetime.now(timezone.utc)
    recommendation = plan.recommendation.value if plan.recommendation else None

    if recommendation in ['CREATE_PAYMENT_LINK', 'OFFER_DISCOUNT']:
        try:
            final_amount = case_data.amount
            discount_cost = 0.0

            if recommendation == 'OFFER_DISCOUNT' and plan.discount_requested_pct and plan.discount_requested_pct > 0:
                discount_multiplier = 1.0 - (plan.discount_requested_pct / 100.0)
                final_amount = round(case_data.amount * discount_multiplier)
                discount_cost = case_data.amount - final_amount

            # Convert to paise
            amount_in_paise = int(final_amount * 100)

            # Create payment link synchronously via razorpay client (which is blocking, but fast enough for this demo, we could wrap in asyncio.to_thread if needed)
            payment_link = razorpay_client.payment_link.create({
                "amount": amount_in_paise,
                "currency": case_data.currency or 'INR',
                "description": f"RevivePay Recovery - Case {case_data.id}",
                "customer": {
                    "name": f"Customer {case_data.customer_id}",
                    "contact": "+919876543210"
                },
                "notify": {
                    "sms": False,
                    "email": False
                },
                "reminder_enable": False,
                "notes": {
                    "case_id": str(case_data.id),
                    "original_amount": case_data.amount,
                    "discount_pct": plan.discount_requested_pct or 0.0,
                    "source": "revivepay"
                }
            })

            return {
                "execution": Execution(
                    executor_type=ExecutorType.RAZORPAY,
                    action_taken=recommendation,
                    result=ExecutionResult.SUCCESS,
                    razorpay_reference=payment_link['id'],
                    payment_link_status=PaymentLinkStatus.CREATED,
                    timestamp=now
                ),
                "status": CaseStatus.AWAITING_PAYMENT,
                "contact_count": case_data.contact_count + 1,
                "last_contacted_at": now,
                "discount_cost": discount_cost,
                "contact_cost": (case_data.contact_count + 1) * 2.0,
                "payment_link_url": payment_link.get('short_url')
            }
        except Exception as e:
            logger.error(f"[RazorpayExecutor] Error creating payment link for case {case_data.id}: {str(e)}")
            return {
                "execution": Execution(
                    executor_type=ExecutorType.RAZORPAY,
                    action_taken=recommendation,
                    result=ExecutionResult.FAILURE,
                    razorpay_reference=None,
                    payment_link_status=None,
                    timestamp=now
                ),
                "status": CaseStatus.EXECUTION_FAILED
            }

    if recommendation == 'SEND_REMINDER':
        logger.info(f"[RazorpayExecutor] Simulated reminder for case {case_data.id}")
        return {
            "execution": Execution(
                executor_type=ExecutorType.RAZORPAY,
                action_taken='SEND_REMINDER',
                result=ExecutionResult.SUCCESS,
                razorpay_reference=None,
                payment_link_status=None,
                timestamp=now
            ),
            "status": CaseStatus.LINK_CREATED,
            "contact_count": case_data.contact_count + 1,
            "last_contacted_at": now,
            "contact_cost": (case_data.contact_count + 1) * 2.0
        }

    return {
        "execution": Execution(
            executor_type=ExecutorType.RAZORPAY,
            action_taken=recommendation,
            result=ExecutionResult.NOT_ATTEMPTED,
            razorpay_reference=None,
            payment_link_status=None,
            timestamp=now
        ),
        "status": CaseStatus.STOPPED_SAFELY if recommendation == 'DO_NOT_CONTACT' else CaseStatus.NEEDS_MERCHANT_APPROVAL
    }
